### DISCUSSION TITLE
pressaction.lua 的替代品

### DISCUSSION THREAD

**Discussion by natural-harmonia-gropius** [dyphire/mpv-config#11] [Ideas]
https://github.com/Natural-Harmonia-Gropius/mpv_config/blob/main/scripts/on.lua
```ini
SPACE               cycle pause                         #@click
SPACE               no-osd set speed 4; set pause no    #@press
SPACE               ignore                              #@release
```
这个脚本搭配上面的 input.conf 应该能够完全代替 pressaction.lua 了。如果测试下来没有问题的话，这个脚本稳定后我将会弃用 pressaction.lua

讨论版：https://github.com/tomasklaen/uosc/discussions/307

**Comment by dyphire** [comment]
pressaction.lua 的功能实现感觉已经很好了（主要是使用频率非常低
不过新脚本的实现方式明显增加了扩展性，尤其是把键位绑定统一放入 input.conf 中管理的想法我很喜欢
等脚本完成后（指正式命名）我应该也会弃用 pressaction.lua

**Comment by dyphire** [comment]
有一个问题，如果其他脚本使用了`mp.add_forced_key_binding`绑定了同键位，是否会产生冲突？
`mp.add_key_binding`呢？看上面 uosc 讨论版里的描述这个似乎不成问题

**Reply by natural-harmonia-gropius** [reply]
```ini
MBTN_LEFT           show-text pressed; script-binding uosc/menu           #@press
MBTN_LEFT           show-text released                                    #@release
```
按压可以触发菜单，松开也会触发。菜单点击事件正常（此时被uosc接管，这个脚本失效），菜单消失，这个脚本重新生效。

在启动时调用 `mp.add_forced_key_binding` 的脚本就不知道了，看运气谁后执行谁生效。

**Reply by natural-harmonia-gropius** [reply]
`mp.add_key_binding` 在我的测试里连 input.conf 都覆盖不掉，后面再试试

**Reply by dyphire** [reply]
> `mp.add_key_binding` 在我的测试里连 input.conf 都覆盖不掉，后面再试试

理论上不禁用--input-default-bindings 的话`mp.add_key_binding`会覆盖 input.conf 的同键位绑定

提到`mp.add_forced_key_binding` 是我感觉到它会产生冲突

**Reply by natural-harmonia-gropius** [reply]
我加了 `no-input-default-bindings` ，怪不得
add_forced_key_binding 确实会冲突，理论上都能靠写 script-binding xxx #@click 解决，没有 #@ 的行不受影响

**Comment by dyphire** [comment]
在测试使用中我发现了一个问题：当我将单击的键位项指定到 `#!` 或 `#menu` 后，如果不指定同键位的 `#@click` 则 inputevent.lua 脚本功能将失效。

这种重复键位功能绑定的处理方式可能不是很好
一种可能的解决方式：在脚本内部匹配已绑定 `#@` 的同键位的 `#!` 和 `#menu` 并视作 `#@click` 处理

**Reply by natural-harmonia-gropius** [reply]
功能上可以这样写... 
![image](https://user-images.githubusercontent.com/50797982/195001239-1cedc3f7-300e-49b9-b095-28b59b9598bc.png)
![image](https://user-images.githubusercontent.com/50797982/195001261-87340e24-0504-4658-932c-21d6de71183d.png)

在解析 input.conf 的步骤里再靠 # 二次分割应该就行了
![image](https://user-images.githubusercontent.com/50797982/195001770-27a2792b-543c-4e12-b658-4bbc1c53d6fa.png)

**Reply by dyphire** [reply]
这种方式渲染的菜单有点怪，感觉还不如上面说的重复键位 `#@click` 功能绑定优雅，暂时先这么使用吧

**Reply by natural-harmonia-gropius** [reply]
我自己这边改完了给 uosc 推个 pr ，应该允许普通注释的

**Reply by natural-harmonia-gropius** [reply]
两边的pr
https://github.com/tomasklaen/uosc/pull/315
https://github.com/Natural-Harmonia-Gropius/InputEvent/pull/12

**Reply by dyphire** [reply]
嗯，看起来不错

**Reply by hooke007** [reply]
一刀切的感觉不是太好，无法通过右侧的按钮提示区分绑定到同一键位的不同触发模式。忘了的话还是得打开input.conf查

**Reply by natural-harmonia-gropius** [reply]
明天我再另外再提交pr吧。
不过长什么样比较好呢，我现在想像这样给按键加个框![image](https://user-images.githubusercontent.com/50797982/195140839-25a99b21-6428-46b3-81a6-3a297cc7fd96.png)
双击单击之类的又要怎么表示比较好？

**Reply by dyphire** [reply]
不清楚是否有简洁的方案去实现这种功能提示，我担心这会让菜单列变得臃肿（更容易触发边界溢出
而且如果对 `#@` 采用针对性识别处理的话我担心这种 pr 不会被合并，只能维护自己的分支（也不是不行

**Reply by natural-harmonia-gropius** [reply]
有什么软件有区分单击双击长按的提示的？
在怎么优雅的做之前，我想不出来怎么做

**Reply by dyphire** [reply]
no idea :(

**Reply by natural-harmonia-gropius** [reply]
😶‍🌫️看着好怪 https://github.com/Natural-Harmonia-Gropius/InputEvent/pull/17

**Reply by dyphire** [reply]
确实好怪，需要特殊化处理。但对如何良好地区分键位动作这点上毫无头绪（

**Comment by dyphire** [comment]
https://github.com/dyphire/mpv-config/commit/d1256f7285ceba2256ebfc05c03233bad785691c 本配置正式启用 inputevent.lua 脚本并弃用 pressaction.lua 脚本

**Reply by zhongfly** [reply]
能帮我测试下我的修改有没有bug吗？
https://github.com/zhongfly/InputEvent/tree/conditional-keybind

* 支持外部键位配置（写法与input.conf完全相同），启用后**只从外部读取增强的键位绑定**，将input.conf中的绑定视为最低优先级的click事件（并忽略其中注释里InputEvent的绑定语句），不然处理外部配置与input.conf之间的优先级是大问题
inputevent.conf
```ini
#enable external config
enable_external_config=yes

#external config file path
external_config=~~/script-opts/inputevent_key.conf
```

* 支持同一种动作根据条件触发不同的命令
条件的语法等同于profile-cond（自动profile的条件语句）
即脚本语法升级为`#@动作|条件`（原先的`#@动作`视为无条件执行命令）
命令执行的优先级是从上到下按顺序，执行第一满足条件的命令（是不是应该改成从下到上？）
示例，在外部配置中写入以下内容
```
SPACE           expand-properties no-osd set speed 3; set pause no     #@press
RIGHT           playlist-next                                          #@click|pause and time_remaining < 2
RIGHT           seek 10                                                #@click #覆盖input.conf的设置
```
同时input.conf中为：
```
SPACE      cycle pause #不需要写@click这样的语句
RIGHT      seek  2
```
效果：
单击空格切换暂停/播放，长按3倍速，松开恢复原状态
剩余时间小于2s且暂停时，单击右方向键直接下一个文件；其余情况依然快进10s

**Reply by dyphire** [reply]
修改版看上去不错，需要测试的应该是条件特性的性能和适应性上的表现

**Reply by zhongfly** [reply]
> 修改版看上去不错，需要测试的应该是条件特性的性能和适应性上的表现

计算条件的性能应该不成问题，基本上是复制自动profile的那个内置脚本(我是没想到这功能原来是内置lua脚本做到的)

**Reply by zhongfly** [reply]
https://github.com/zhongfly/InputEvent/tree/property-expansion 合并了原作者对属性扩展增强的pr，支持`add speed ${?speed==4.00:0}${!speed==4.00:0.1}`这样mpv原本会报错的属性扩展（add命令不支持）

**Reply by dyphire** [reply]
测试了下修改版inputevent.lua，整体的体验不错。提两个小建议：

1. 当启用外部配置的特性时（默认启用）假如外部配置不存在应尝试读取input.conf而不是打印错误并静默失败，可以确保向下兼容
2. 我个人喜欢在input.conf中统一管理静态键位，即使inputevent.lua脚本使用外部配置也同样是对input.conf的增强，默认路径应该放在input.conf旁而不是脚本配置目录中，不便于静态键位管理

**Reply by zhongfly** [reply]
> 测试了下修改版inputevent.lua，整体的体验不错。提两个小建议：
> 
> 1. 当启用外部配置的特性时（默认启用）假如外部配置不存在应尝试读取input.conf而不是打印错误并静默失败，可以确保向下兼容
> 2. 我个人喜欢在input.conf中统一管理静态键位，即使inputevent.lua脚本使用外部配置也同样是对input.conf的增强，默认路径应该放在input.conf旁而不是脚本配置目录中，不便于静态键位管理

之后变成默认不启用外部配置
由于是脚本配置，还是按一般习惯放在脚本配置目录，有需求的自己改路径

**Reply by dyphire** [reply]
> 之后变成默认不启用外部配置

这并没有真正解决问题，就像我说的当启用时外部配置不存在应该尝试读取input.conf中可能存在的配置而不是静默无效，当外部配置存在保持现有逻辑

> 由于是脚本配置，还是按一般习惯放在脚本配置目录，有需求的自己改路径

这是个人偏好问题，倒不是很必要

**Reply by zhongfly** [reply]
> > 之后变成默认不启用外部配置
> 
> 这并没有真正解决问题，就像我说的当启用时外部配置不存在应该尝试读取input.conf中可能存在的配置而不是静默无效，当外部配置存在保持现有逻辑

在默认不启用时，用户只可能自己在设置里显式启用，这说明他确定要使用外部配置、不使用input.conf。如果开启外部配置却还是读取input.conf是不合逻辑的。没读取到，说明他忘记或者写错路径了，那么用户需要纠正设置

**Reply by dyphire** [reply]
嗯，这也是有道理的

**Reply by dyphire** [reply]
https://github.com/dyphire/mpv-config/commit/abd5eae49cb9ac290c87e2a4a4d562bc0126a5ed 切换至修改版inputevent.lua进行测试

> 命令执行的优先级是从上到下按顺序，执行第一满足条件的命令（是不是应该改成从下到上？）

个人觉得从下到上的优先级会更好些，一般的写法习惯是先写常规动作再写限定动作，目前的逻辑和此相反，改下会比较好
然后更新下readme后感觉可以尝试提交pr了

**Reply by zhongfly** [reply]
改完了，文档实在不想写了

**Reply by zhongfly** [reply]
把条件命令剥离出来作为单独的脚本，这样就不用修改版了 https://github.com/zhongfly/command_cond
依然是支持设置从哪读取配置文件
通过script-message-to来掉用
优先级我还是觉得改成从上到下更好，更符合直觉，即最下面的是“兜底”。我自己感觉涉及规则类的都是从上到下匹配的(广告屏蔽，代理规则等等)

**Comment by dyphire** [comment]
我注意到了和 https://github.com/natural-harmonia-gropius/input-event/issues/34 相似的问题，inputevent.lua 的键位劫持会破坏某些脚本的动态键位。
复现步骤：使用主仓库配置，inputevent.lua 使用原版或修改版都可以复现；反复打开一些菜单类脚本测试动态键位是否工作。
测试视频中可以看到最后 uosc 的部分动态键位不再生效（再继续反复打开菜单类脚本的话 uosc 的全部动态键位都会失效

https://user-images.githubusercontent.com/61936050/232290085-b291424a-fd75-4510-9cde-1ceed6ea1c5f.mp4

移除 inputevent.lua 脚本或 input*.conf 中的 `#@event` 相关键位动作绑定将修复此问题；移除主仓库中任一菜单类脚本也可暂时避免此问题
最初是在此提交 https://github.com/dyphire/mpv-config/commit/78012f64db10ea8e3e5a0d5073b4a570f02e2971 时期发现了动态键位绑定的一些问题，不过当时并没有定位到 inputevent.lua 上面

**Reply by natural-harmonia-gropius** [reply]
我感觉是mpv的问题，脚本这边我找不出问题

**Reply by dyphire** [reply]
演示视频中触发的错误和uosc的鼠标动态键位绑定方式有关，有点与众不同，通过[此提交内容](https://github.com/dyphire/mpv-config/commit/334d1b45d69e51b6c0a0e08863c4625e534859d9#diff-8b51ce5a77d2bead2dc85b1cbdd70197ea412fbfbc06051ca03fa21582121244)可以修复

https://github.com/dyphire/mpv-config/commit/78012f64db10ea8e3e5a0d5073b4a570f02e2971 中发现的键位争用相关bug的触发原因就不明了，应该是mpv的问题
